File: request_sig_via_email/request_sig_via_email.m
//
//  API Walkthrough 4 - Request Signature on a Document (Objective-C)
//
//  To run this sample:
//      1.  Copy the below code into your iOS project
//      2.  Enter email, password, integrator key, name, and document name into variables and save
//      3.  Compile and Run
//

// Enter your info:
NSString *email = @"<#email#>";
NSString *password = @"<#password#>";
NSString *integratorKey = @"<#integratorKey#>";
NSString *name = @"<#recipientName#>";

// provide filename for document and make sure you copy file with same name into app directory
NSString *documentName = @"<#test.pdf#>";


- (void)requestSignatureOnDocument
{

    ///////////////////////////////////////////////////////////////////////////////////////
    // STEP 1 - Login (retrieves accountId and baseUrl)
    ///////////////////////////////////////////////////////////////////////////////////////
    
    // endpoint for the Login api
    NSString *loginURL = @"https://demo.docusign.net/restapi/v2/login_information";
    
    // initialize a request object with the following url, method, and body (no body for login api call)
    NSMutableURLRequest *loginRequest = [self initializeRequest:loginURL setMethod:@"GET" setBody:nil];
    
    // add content-type and authorization headers
    [self addRequestHeaders:loginRequest];
    
    //--> make an asynchronous web request
    [NSURLConnection sendAsynchronousRequest:loginRequest queue:[NSOperationQueue mainQueue] completionHandler:^(NSURLResponse *loginResponse, NSData *loginData, NSError *loginError) {
        
        // create NSDictionary out of JSON data that is returned
        NSDictionary *responseDictionary = [self getHttpResponse:loginRequest response:loginResponse data:loginData error:loginError];
        if( responseDictionary == nil)
            return;
        
        // parse the accountId and baseUrl from the response and use in the next request
        NSString *accountId = responseDictionary[@"loginAccounts"][0][@"accountId"];
        NSString *baseUrl = responseDictionary[@"loginAccounts"][0][@"baseUrl"];
        
        //--- display results
        NSLog(@"\naccountId = %@\nbaseUrl = %@\n", accountId, baseUrl);
        
        ///////////////////////////////////////////////////////////////////////////////////////
        // STEP 2 - Request Signature on Document
        ///////////////////////////////////////////////////////////////////////////////////////
        
        // append "/envelopes" URI to your baseUrl and use as endpoint for signature request call
        NSString *signatureURL = [NSMutableString stringWithFormat:@"%@/envelopes",baseUrl];
        
        // construct a JSON formatted signature request body (multi-line for readability)
        NSDictionary *signatureRequestData =
        @{@"accountId": accountId,
          @"emailSubject" : @"Signature Request on Document API call",
          @"emailBlurb" : @"email body",
          @"documents" : [NSArray arrayWithObjects: @{@"documentId":@"1", @"name": documentName}, nil ],
          @"recipients" : @{ @"signers": [NSArray arrayWithObjects:
                                          @{@"email": email,
                                            @"name": name,
                                            @"recipientId": @"1",
                                            @"tabs": @{ @"signHereTabs": [NSArray arrayWithObjects:
                                                                          @{@"xPosition": @"100",
                                                                            @"yPosition": @"100",
                                                                            @"documentId": @"1",
                                                                            @"pageNumber": @"1"}, nil ]}}, nil ] },
          @"status" : @"sent"
          };
        
        // convert dictionary object to JSON formatted string
        NSString *sigRequestDataString = [self jsonStringFromObject:signatureRequestData];
        
        // read document bytes and place in the request
        NSString *appDirectory = [[[NSBundle mainBundle] bundlePath] stringByDeletingLastPathComponent];
        NSString *fullFilePath = [NSString stringWithFormat:@"%@/%@", appDirectory, documentName];
        
        // make sure document exists in local app directory
        if( !([[NSFileManager defaultManager] fileExistsAtPath: fullFilePath]) )
        {
            NSLog(@"\n\nCould not find test document \"%@\".  Please copy this file into following app directory: %@\n\n", documentName, fullFilePath);
            return;
        }
        
        // use an NSData object to store the document bytes
        NSData *filedata = [NSData dataWithContentsOfFile:fullFilePath];
        
        // create the boundary separated request body...
        NSMutableData *body = [NSMutableData data];
        [body appendData:[[NSString stringWithFormat:
                           @"\r\n"
                           "\r\n"
                           "--AAA\r\n"
                           "Content-Type: application/json\r\n"
                           "Content-Disposition: form-data\r\n"
                           "\r\n"
                           "%@\r\n"
                           "--AAA\r\n"
                           "Content-Type: application/pdf\r\n"
                           "Content-Disposition: file; filename=\"%@\"; documentid=1; fileExtension=\"pdf\" \r\n"
                           "\r\n",
                           sigRequestDataString, documentName] dataUsingEncoding:NSUTF8StringEncoding]];
        
        // next append the document bytes
        [body appendData:filedata];
        
        // append closing boundary and CRLFs to the request
        [body appendData:[[NSString stringWithFormat:
                           @"\r\n"
                           "--AAA--\r\n"
                           "\r\n"] dataUsingEncoding:NSUTF8StringEncoding]];
        
        // initialize a request object with the following url, method, and body (no body for login api call)
        NSMutableURLRequest *signatureRequest = [self initializeRequest:signatureURL setMethod:@"POST" setBody:body];
        
        // add content-type and authorization headers
        [self addRequestHeaders:signatureRequest];
        
        // this is the only call where we set the content type to multipart/form-data and set a request boundary:
        [signatureRequest setValue:@"multipart/form-data; boundary=AAA" forHTTPHeaderField:@"Content-Type"];
        
        //*** make the signature request!
        [NSURLConnection sendAsynchronousRequest:signatureRequest queue:[NSOperationQueue mainQueue] completionHandler:^(NSURLResponse *envelopeResponse, NSData *envelopeData, NSError *envelopeError) {
            
            NSError *jsonError = nil;
            NSDictionary *responseDictionary = [NSJSONSerialization JSONObjectWithData:envelopeData options:kNilOptions error:&jsonError];
            
            //--- display results
            NSLog(@"Envelope Sent!  Response is: %@\n", responseDictionary);
        }];
    }];
}

//***********************************************************************************************
// --- HELPER FUNCTIONS ---
//***********************************************************************************************
- (NSMutableURLRequest *) initializeRequest:(NSString *) url setMethod:(NSString *) method setBody:(NSData *) body
{
    // create a request using the passed parameters for url, method, and body
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
    [request setHTTPMethod:method];
    [request setHTTPBody:body];
    [request setURL:[NSURL URLWithString:url]];
    return request;
}
/////////////////////////////////////////////////////////////////////////////////////////////////
- (void) addRequestHeaders:(NSMutableURLRequest *) request
{
    // set JSON formatted X-DocuSign-Authentication header (XML format also accepted)
    NSDictionary *authenticationHeader = @{ @"Username": email, @"Password" : password, @"IntegratorKey" : integratorKey };
    
    // add the auth header to the request object
    [request setValue:[self jsonStringFromObject:authenticationHeader] forHTTPHeaderField:@"X-DocuSign-Authentication"];
    
    // also set the Content-Type header (other accepted type is application/xml)
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
}
/////////////////////////////////////////////////////////////////////////////////////////////////
- (NSDictionary *) getHttpResponse:(NSMutableURLRequest *) request response:(NSURLResponse *) response data:(NSData *) data error:(NSError *) error
{
    if (error) {
        NSLog(@"Error sending request %@\n. Got Response %@\n Error is: %@\n", request, response, error);
        return nil;
    }
    // we use NSJSONSerialization to parse the JSON formatted response
    NSError *jsonError = nil;
    NSDictionary *resp = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&jsonError];
    return resp;
}
/////////////////////////////////////////////////////////////////////////////////////////////////
- (NSString *)jsonStringFromObject:(id)object {
    NSString *string = [[NSString alloc] initWithData:[NSJSONSerialization dataWithJSONObject:object options:0 error:nil] encoding:NSUTF8StringEncoding];
    return string;
}

File: request_sig_via_email/request_sig_via_email.js
// Request Signature on a Document (Node.js)

// To run this sample
//  1. Copy the file to your local machine and give .js extension (i.e. example.js)
//  2. Change "***" to appropriate values
//  3. Install docusign-esign, async, and fs packages
//     npm install docusign-esign
//     npm install async
//     npm install fs
//  4. Ensure sure 'blank.pdf' exists in the same directory, or create a blank pdf yourself
//  5. Execute
//     node example.js 
//

var docusign = require('docusign-esign'),
	async = require('async'),
	fs = require('fs'),
	path = require('path');

var integratorKey = '***',	// Integrator Key associated with your DocuSign Integration
	email = '***',			// Email for your DocuSign Account
	password = '***',		// Password for your DocuSign Account
	recipientName = '***',	// Recipient's Full Name
	recipientEmail = '***', // Recipient's Email
	docusignEnv = 'demo',	// DocuSign Environment generally demo for testing purposes ('www' == production)
	fileToSign = 'blank.pdf',
	baseUrl = 'https://' + docusignEnv + '.docusign.net/restapi';


// relative path to PDF (can also be many other filetypes) 

async.waterfall(
  [
	/////////////////////////////////////////////////////////////////////////////////////
	// Step 1: Login (used to retrieve your accountId and account baseUrl)
	/////////////////////////////////////////////////////////////////////////////////////

	function login(next) {

		// initialize the api client
		var apiClient = new docusign.ApiClient();
		apiClient.setBasePath(baseUrl);

		// create JSON formatted auth header
		var creds = JSON.stringify({
		  Username: email,
		  Password: password,
		  IntegratorKey: integratorKey
		});
		apiClient.addDefaultHeader('X-DocuSign-Authentication', creds);

		// assign api client to the Configuration object
		docusign.Configuration.default.setDefaultApiClient(apiClient);

		// login call available off the AuthenticationApi
		var authApi = new docusign.AuthenticationApi();

		// login has some optional parameters we can set
		var loginOps = new authApi.LoginOptions();
		loginOps.setApiPassword('true');
		loginOps.setIncludeAccountIdGuid('true');
		authApi.login(loginOps, function (err, loginInfo, response) {
			if (err) {
				console.error(err.response ? err.response.error : err);
				return;
			}
			console.log('Login Info');
			console.log(JSON.stringify(loginInfo,null,2));
			if (loginInfo) {
				// list of user account(s)
				// note that a given user may be a member of multiple accounts
				var loginAccounts = loginInfo.getLoginAccounts();
				console.log('LoginInformation: ' + JSON.stringify(loginAccounts));
				next(null, loginAccounts);
			}
		});
	},
	
	/////////////////////////////////////////////////////////////////////////////////////
	// Step 2: Request Signature on a PDF Document
	/////////////////////////////////////////////////////////////////////////////////////

	function requestSignature(loginAccounts, next){
		console.log('requestSignature');
		// create a byte array that will hold our document bytes
		var fileBytes = null;
		try {
			// read file from a local directory
			fileBytes = fs.readFileSync(path.resolve([__filename, '..', fileToSign].join('/')));
		} catch (ex) {
			// handle error
			console.log('Exception: ' + ex);
			return;
		}

		// create an envelope that will store the document(s), field(s), and recipient(s)
		var envDef = new docusign.EnvelopeDefinition();
		envDef.setEmailSubject('Please sign this document sent from the DocuSign Node SDK)');

		// add a document to the envelope
		var doc = new docusign.Document();
		var base64Doc = new Buffer(fileBytes).toString('base64');
		doc.setDocumentBase64(base64Doc);
		doc.setName('DocuSign API - Signature Request on Document Call'); // can be different from actual file name
		doc.setDocumentId('1'); // hardcode so we can easily refer to this document later

		var docs = [];
		docs.push(doc);
		envDef.setDocuments(docs);

		// add a recipient to sign the document, identified by name and email we used above
		var signer = new docusign.Signer();
		signer.setEmail(recipientEmail);
		signer.setName(recipientName);
		signer.setRecipientId('1');

		// create a signHere tab somewhere on the document for the signer to sign
		// default unit of measurement is pixels, can be mms, cms, inches also
		var signHere = new docusign.SignHere();
		signHere.setDocumentId('1');
		signHere.setPageNumber('1');
		signHere.setRecipientId('1');
		signHere.setXPosition('100');
		signHere.setYPosition('100');

		// can have multiple tabs, so need to add to envelope as a single element list
		var signHereTabs = [];
		signHereTabs.push(signHere);
		var tabs = new docusign.Tabs();
		tabs.setSignHereTabs(signHereTabs);
		signer.setTabs(tabs);

		// add recipients (in this case a single signer) to the envelope
		envDef.setRecipients(new docusign.Recipients());
		envDef.getRecipients().setSigners([]);
		envDef.getRecipients().getSigners().push(signer);

		// send the envelope by setting |status| to "sent". To save as a draft set to "created"
		// - note that the envelope will only be 'sent' when it reaches the DocuSign server with the 'sent' status (not in the following call)
		envDef.setStatus('sent');

		// use the |accountId| we retrieved through the Login API to create the Envelope
		var loginAccount = new docusign.LoginAccount();
		loginAccount = loginAccounts[0];
		var accountId = loginAccount.accountId;

		// instantiate a new EnvelopesApi object
		var envelopesApi = new docusign.EnvelopesApi();

		// call the createEnvelope() API
		envelopesApi.createEnvelope(accountId, envDef, null, function (error, envelopeSummary, response) {
			if (error) {
				console.error('Error: ' + error);
				return;
			}

			if (envelopeSummary) {
				console.log('EnvelopeSummary: ' + JSON.stringify(envelopeSummary,null,2));
			}
		});
	}

]);
